File: packages/python/src/slybrowser/manifest.py
"""Signed release-manifest and browser-artifact verification."""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Mapping

from .canonical import canonical_json, decode_base64url
from .errors import ArtifactError, ManifestError

MAX_MANIFEST_BYTES = 1024 * 1024


def _numeric_version(value: str) -> tuple[int, ...]:
    import re
    if not isinstance(value, str) or not re.fullmatch(r"\d+(?:\.\d+){0,7}", value):
        raise ManifestError("SDK version is invalid", code="sdk_version_invalid")
    return tuple(int(part) for part in value.split("."))


def is_sdk_compatible(value: str, version: str) -> bool:
    import re
    current = _numeric_version(version)
    tokens = value.strip().split()
    if not tokens:
        raise ManifestError("Manifest SDK compatibility is invalid", code="manifest_invalid")
    for token in tokens:
        if token.startswith("^"):
            target = _numeric_version(token[1:])
            major = target[0] if len(target) > 0 else 0
            minor = target[1] if len(target) > 1 else 0
            patch = target[2] if len(target) > 2 else 0
            if major > 0:
                upper = (major + 1, 0, 0)
            elif minor > 0:
                upper = (0, minor + 1, 0)
            else:
                upper = (0, 0, patch + 1)
            length = max(len(current), len(target), len(upper))
            left = current + (0,) * (length - len(current))
            lower = target + (0,) * (length - len(target))
            right = upper + (0,) * (length - len(upper))
            if not (left >= lower and left < right):
                return False
            continue
        match = re.fullmatch(r"(>=|<=|>|<|=)?(\d+(?:\.\d+){0,7})", token)
        if not match:
            raise ManifestError("Manifest SDK compatibility is invalid", code="manifest_invalid")
        target = _numeric_version(match.group(2))
        length = max(len(current), len(target))
        left = current + (0,) * (length - len(current))
        right = target + (0,) * (length - len(target))
        operator = match.group(1) or "="
        if operator == ">=" and not left >= right:
            return False
        if operator == "<=" and not left <= right:
            return False
        if operator == ">" and not left > right:
            return False
        if operator == "<" and not left < right:
            return False
        if operator == "=" and not left == right:
            return False
    return True


@dataclass(frozen=True, slots=True)
class ReleasePrivateModule:
    path: str
    sha256: str
    size: int
    abi: str


@dataclass(frozen=True, slots=True)
class ReleaseResourceFile:
    path: str
    sha256: str
    size: int


@dataclass(frozen=True, slots=True)
class ReleaseCodeSignature:
    scheme: str
    subject: str
    certificateSha256: str
    timestampRequired: bool


@dataclass(frozen=True, slots=True)
class ReleaseArtifact:
    platform: str
    arch: str
    url: str
    sha256: str
    size: int
    archiveFormat: str
    browserExecutable: str
    driverExecutable: str
    browserSha256: str
    driverSha256: str
    privateModules: tuple[ReleasePrivateModule, ...]
    resources: tuple[ReleaseResourceFile, ...]
    codeSignature: ReleaseCodeSignature | None = None


@dataclass(frozen=True, slots=True)
class ReleaseManifest:
    browser_version: str
    sdk_compatibility: str
    status: str
    artifacts: tuple[ReleaseArtifact, ...]
    evidence: Mapping[str, Any] | None
    signing_key_id: str
    raw: Mapping[str, Any]

    def select(self, platform: str, arch: str) -> ReleaseArtifact:
        matches = [item for item in self.artifacts if item.platform == platform and item.arch == arch]
        if len(matches) != 1:
            available = ", ".join(sorted(f"{item.platform}/{item.arch}" for item in self.artifacts))
            raise ManifestError(
                f"No signed artifact supports {platform}/{arch}; available targets: {available}",
                code="artifact_not_found",
            )
        return matches[0]


def verify_release_manifest(
    raw: bytes | str | Mapping[str, Any],
    *,
    trusted_keys: Mapping[str, bytes],
) -> ReleaseManifest:
    document = _read_document(raw)
    signature = document.get("signature")
    if not isinstance(signature, dict) or set(signature) != {"algorithm", "keyId", "value"}:
        raise ManifestError("Manifest signature block is invalid", code="manifest_invalid_signature")
    if signature.get("algorithm") != "ed25519":
        raise ManifestError("Manifest signature algorithm is unsupported", code="manifest_algorithm_unsupported")
    key_id = signature.get("keyId")
    if not isinstance(key_id, str) or key_id not in trusted_keys:
        raise ManifestError("Manifest signing key is unknown", code="manifest_key_unknown")
    public_key = trusted_keys[key_id]
    if len(public_key) != 32:
        raise ManifestError("Manifest public key is invalid", code="manifest_key_invalid")
    try:
        signature_bytes = decode_base64url(signature["value"], max_bytes=64)
    except (KeyError, ValueError) as exc:
        raise ManifestError("Manifest signature encoding is invalid", code="manifest_invalid_signature") from exc
    if len(signature_bytes) != 64:
        raise ManifestError("Manifest signature length is invalid", code="manifest_invalid_signature")

    payload = dict(document)
    del payload["signature"]
    try:
        from cryptography.exceptions import InvalidSignature
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
    except ImportError as exc:
        raise ManifestError("Ed25519 support is unavailable", code="manifest_crypto_unavailable") from exc
    try:
        Ed25519PublicKey.from_public_bytes(public_key).verify(signature_bytes, canonical_json(payload))
    except InvalidSignature as exc:
        raise ManifestError("Manifest signature is invalid", code="manifest_invalid_signature") from exc

    if document.get("schemaVersion") != 1:
        raise ManifestError("Manifest schema is unsupported", code="manifest_schema_unsupported")
    browser_version = document.get("browserVersion")
    sdk_compatibility = document.get("sdkCompatibility")
    artifacts_document = document.get("artifacts")
    if not isinstance(browser_version, str) or not browser_version:
        raise ManifestError("Manifest browser version is invalid", code="manifest_invalid")
    if not isinstance(sdk_compatibility, str) or not sdk_compatibility:
        raise ManifestError("Manifest SDK compatibility is invalid", code="manifest_invalid")
    status = document.get("status")
    if status not in {"available", "revoked"}:
        raise ManifestError("Manifest release status is invalid", code="manifest_invalid")
    if not isinstance(artifacts_document, list) or not artifacts_document:
        raise ManifestError("Manifest artifacts are invalid", code="manifest_invalid")
    artifacts = tuple(_parse_artifact(item) for item in artifacts_document)
    if len({(item.platform, item.arch) for item in artifacts}) != len(artifacts):
        raise ManifestError("Manifest has duplicate platform artifacts", code="manifest_duplicate_artifact")
    evidence = None if "evidence" not in document else _parse_evidence(document.get("evidence"))
    return ReleaseManifest(browser_version, sdk_compatibility, status, artifacts, evidence, key_id, document)


def _parse_evidence_artifact(value: object, media_type: str) -> dict[str, Any]:
    if not isinstance(value, dict) or set(value) != {"url", "sha256", "size", "mediaType"}:
        raise ManifestError("Release evidence is invalid", code="manifest_evidence_invalid")
    valid = (
        isinstance(value["url"], str) and value["url"].startswith("https://")
        and isinstance(value["sha256"], str) and len(value["sha256"]) == 64
        and all(character in "0123456789abcdef" for character in value["sha256"])
        and isinstance(value["size"], int) and not isinstance(value["size"], bool) and value["size"] > 0
        and value["mediaType"] == media_type
    )
    if not valid:
        raise ManifestError("Release evidence is invalid", code="manifest_evidence_invalid")
    return dict(value)


def _parse_evidence(value: object) -> dict[str, Any]:
    required = {"sbom", "provenance", "chromiumPatchInventory", "sourceBoundary"}
    if not isinstance(value, dict):
        raise ManifestError("Release evidence is missing", code="manifest_evidence_missing")
    if set(value) != required:
        raise ManifestError("Release evidence fields are invalid", code="manifest_evidence_invalid")
    boundary = value["sourceBoundary"]
    if not isinstance(boundary, dict) or boundary != {
        "sdk": "open-source",
        "chromiumPatches": "inventory-and-approved-patches",
        "proprietaryCore": "private",
    }:
        raise ManifestError("Source boundary is invalid", code="manifest_evidence_invalid")
    return {
        "sbom": _parse_evidence_artifact(value["sbom"], "application/vnd.cyclonedx+json"),
        "provenance": _parse_evidence_artifact(value["provenance"], "application/vnd.in-toto+json"),
        "chromiumPatchInventory": _parse_evidence_artifact(
            value["chromiumPatchInventory"],
            "application/vnd.slybrowser.chromium-patch-inventory+json",
        ),
        "sourceBoundary": dict(boundary),
    }


def verify_artifact(path: str | Path, artifact: ReleaseArtifact) -> None:
    artifact_path = Path(path)
    if not artifact_path.is_file():
        raise ArtifactError("Browser artifact is missing", code="artifact_missing")
    stat = artifact_path.stat()
    if stat.st_size != artifact.size:
        raise ArtifactError("Browser artifact size does not match", code="artifact_size_mismatch")
    digest = hashlib.sha256()
    with artifact_path.open("rb") as handle:
        for block in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(block)
    if digest.hexdigest() != artifact.sha256:
        raise ArtifactError("Browser artifact checksum does not match", code="artifact_hash_mismatch")


def _read_document(raw: bytes | str | Mapping[str, Any]) -> dict[str, Any]:
    if isinstance(raw, Mapping):
        return dict(raw)
    value = raw.encode("utf-8") if isinstance(raw, str) else raw
    if not isinstance(value, bytes) or len(value) > MAX_MANIFEST_BYTES:
        raise ManifestError("Manifest exceeds its size limit", code="manifest_too_large")
    try:
        document = json.loads(value)
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ManifestError("Manifest is not valid JSON", code="manifest_invalid") from exc
    if not isinstance(document, dict):
        raise ManifestError("Manifest must be an object", code="manifest_invalid")
    return document


def _parse_artifact(value: Any) -> ReleaseArtifact:
    if not isinstance(value, dict):
        raise ManifestError("Artifact entry must be an object", code="manifest_invalid")
    required = {
        "platform", "arch", "url", "sha256", "size", "archiveFormat",
        "browserExecutable", "driverExecutable", "browserSha256", "driverSha256",
        "privateModules", "resources",
    }
    allowed = required | {"codeSignature"}
    if not required.issubset(set(value)) or not set(value).issubset(allowed):
        raise ManifestError("Artifact fields are invalid", code="manifest_invalid")
    if value["platform"] not in {"windows", "linux", "macos"}:
        raise ManifestError("Artifact platform is invalid", code="manifest_invalid")
    if value["arch"] not in {"x64", "arm64"}:
        raise ManifestError("Artifact architecture is invalid", code="manifest_invalid")
    if not isinstance(value["url"], str) or not value["url"].startswith("https://"):
        raise ManifestError("Artifact URL must use HTTPS", code="manifest_invalid")
    if (
        not isinstance(value["sha256"], str)
        or len(value["sha256"]) != 64
        or any(character not in "0123456789abcdef" for character in value["sha256"])
    ):
        raise ManifestError("Artifact SHA-256 is invalid", code="manifest_invalid")
    if not isinstance(value["size"], int) or isinstance(value["size"], bool) or value["size"] <= 0:
        raise ManifestError("Artifact size is invalid", code="manifest_invalid")
    if value["archiveFormat"] != "zip":
        raise ManifestError("Artifact archive format is invalid", code="manifest_invalid")
    for name in ("browserExecutable", "driverExecutable"):
        if not _safe_relative_path(value[name]):
            raise ManifestError("Artifact runtime path is invalid", code="manifest_invalid")
    for name in ("browserSha256", "driverSha256"):
        if not isinstance(value[name], str) or len(value[name]) != 64 or any(character not in "0123456789abcdef" for character in value[name]):
            raise ManifestError("Artifact runtime hash is invalid", code="manifest_invalid")
    private_modules = _parse_private_modules(value["privateModules"])
    resources = _parse_resources(value["resources"])
    code_signature = _parse_code_signature(value["codeSignature"]) if "codeSignature" in value else None
    return ReleaseArtifact(
        value["platform"],
        value["arch"],
        value["url"],
        value["sha256"],
        value["size"],
        value["archiveFormat"],
        value["browserExecutable"],
        value["driverExecutable"],
        value["browserSha256"],
        value["driverSha256"],
        private_modules,
        resources,
        code_signature,
    )


def _safe_relative_path(value: object) -> bool:
    if not isinstance(value, str) or not value or len(value) > 512:
        return False
    normalized = value.replace("\\", "/")
    return not normalized.startswith("/") and not (len(normalized) >= 2 and normalized[1] == ":") and ".." not in normalized.split("/")


def _hex64(value: object) -> bool:
    return isinstance(value, str) and len(value) == 64 and not any(character not in "0123456789abcdef" for character in value)


def _parse_private_modules(value: object) -> tuple[ReleasePrivateModule, ...]:
    if not isinstance(value, list) or not value:
        raise ManifestError("Artifact private module metadata is invalid", code="manifest_invalid")
    modules: list[ReleasePrivateModule] = []
    for item in value:
        if not isinstance(item, dict) or set(item) != {"path", "sha256", "size", "abi"}:
            raise ManifestError("Artifact private module metadata is invalid", code="manifest_invalid")
        if (
            not _safe_relative_path(item["path"])
            or not _hex64(item["sha256"])
            or not isinstance(item["size"], int)
            or isinstance(item["size"], bool)
            or item["size"] <= 0
            or not isinstance(item["abi"], str)
            or not item["abi"]
            or len(item["abi"]) > 128
        ):
            raise ManifestError("Artifact private module metadata is invalid", code="manifest_invalid")
        modules.append(ReleasePrivateModule(item["path"], item["sha256"], item["size"], item["abi"]))
    return tuple(modules)


def _parse_resources(value: object) -> tuple[ReleaseResourceFile, ...]:
    if not isinstance(value, list) or not value:
        raise ManifestError("Artifact resource metadata is invalid", code="manifest_invalid")
    resources: list[ReleaseResourceFile] = []
    for item in value:
        if not isinstance(item, dict) or set(item) != {"path", "sha256", "size"}:
            raise ManifestError("Artifact resource metadata is invalid", code="manifest_invalid")
        if (
            not _safe_relative_path(item["path"])
            or not _hex64(item["sha256"])
            or not isinstance(item["size"], int)
            or isinstance(item["size"], bool)
            or item["size"] <= 0
        ):
            raise ManifestError("Artifact resource metadata is invalid", code="manifest_invalid")
        resources.append(ReleaseResourceFile(item["path"], item["sha256"], item["size"]))
    return tuple(resources)


def _parse_code_signature(value: object) -> ReleaseCodeSignature:
    if not isinstance(value, dict) or set(value) != {"scheme", "subject", "certificateSha256", "timestampRequired"}:
        raise ManifestError("Artifact code signature metadata is invalid", code="manifest_invalid")
    if (
        value["scheme"] not in {"authenticode", "apple-developer-id", "x509-code-signing"}
        or not isinstance(value["subject"], str)
        or not value["subject"]
        or len(value["subject"]) > 512
        or not _hex64(value["certificateSha256"])
        or not isinstance(value["timestampRequired"], bool)
    ):
        raise ManifestError("Artifact code signature metadata is invalid", code="manifest_invalid")
    return ReleaseCodeSignature(
        value["scheme"],
        value["subject"],
        value["certificateSha256"],
        value["timestampRequired"],
    )

File: packages/node/src/manifest.ts
import { createHash, createPublicKey, verify as verifySignature } from "node:crypto";
import { createReadStream } from "node:fs";
import { stat } from "node:fs/promises";

import { canonicalJson, decodeBase64Url } from "./canonical.js";
import { ArtifactError, ManifestError } from "./errors.js";

const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export interface ReleaseArtifact {
  platform: "windows" | "linux" | "macos";
  arch: "x64" | "arm64";
  url: string;
  sha256: string;
  size: number;
  archiveFormat: "zip";
  browserExecutable: string;
  driverExecutable: string;
  browserSha256: string;
  driverSha256: string;
  privateModules: ReleasePrivateModule[];
  resources: ReleaseResourceFile[];
  codeSignature?: ReleaseCodeSignature;
}

export interface ReleasePrivateModule {
  path: string;
  sha256: string;
  size: number;
  abi: string;
}

export interface ReleaseResourceFile {
  path: string;
  sha256: string;
  size: number;
}

export interface ReleaseCodeSignature {
  scheme: "authenticode" | "apple-developer-id" | "x509-code-signing";
  subject: string;
  certificateSha256: string;
  timestampRequired: boolean;
}

export interface ReleaseEvidenceArtifact {
  url: string;
  sha256: string;
  size: number;
  mediaType: "application/vnd.cyclonedx+json" | "application/vnd.in-toto+json" |
    "application/vnd.slybrowser.chromium-patch-inventory+json";
}

export interface ReleaseEvidence {
  sbom: ReleaseEvidenceArtifact;
  provenance: ReleaseEvidenceArtifact;
  chromiumPatchInventory: ReleaseEvidenceArtifact;
  sourceBoundary: {
    sdk: "open-source";
    chromiumPatches: "inventory-and-approved-patches";
    proprietaryCore: "private";
  };
}

export interface ReleaseManifest {
  schemaVersion: 1;
  browserVersion: string;
  sdkCompatibility: string;
  status: "available" | "revoked";
  artifacts: ReleaseArtifact[];
  evidence?: ReleaseEvidence;
  signature: { algorithm: "ed25519"; keyId: string; value: string };
}

function fail(message: string, code: string): never {
  throw new ManifestError(message, code);
}

function numericVersion(value: string): number[] {
  if (!/^\d+(\.\d+){0,7}$/.test(value)) fail("SDK version is invalid", "sdk_version_invalid");
  return value.split(".").map(Number);
}

function compareNumericVersions(left: string, right: string): number {
  const a = numericVersion(left);
  const b = numericVersion(right);
  for (let index = 0; index < Math.max(a.length, b.length); index += 1) {
    const difference = (a[index] ?? 0) - (b[index] ?? 0);
    if (difference !== 0) return Math.sign(difference);
  }
  return 0;
}

export function isSdkCompatible(range: string, version: string): boolean {
  numericVersion(version);
  const tokens = range.trim().split(/\s+/).filter(Boolean);
  if (!tokens.length) fail("Manifest SDK compatibility is invalid", "manifest_invalid");
  return tokens.every((token) => {
    if (token.startsWith("^")) return isCaretCompatible(token.slice(1), version);
    const match = /^(>=|<=|>|<|=)?(\d+(?:\.\d+){0,7})$/.exec(token);
    if (!match) fail("Manifest SDK compatibility is invalid", "manifest_invalid");
    const comparison = compareNumericVersions(version, match[2]!);
    switch (match[1] ?? "=") {
      case ">=": return comparison >= 0;
      case "<=": return comparison <= 0;
      case ">": return comparison > 0;
      case "<": return comparison < 0;
      default: return comparison === 0;
    }
  });
}

function isCaretCompatible(base: string, version: string): boolean {
  const parts = numericVersion(base);
  const major = parts[0] ?? 0;
  const minor = parts[1] ?? 0;
  const patch = parts[2] ?? 0;
  const upper = major > 0
    ? `${major + 1}.0.0`
    : minor > 0
      ? `0.${minor + 1}.0`
      : `0.0.${patch + 1}`;
  return compareNumericVersions(version, base) >= 0 && compareNumericVersions(version, upper) < 0;
}

export function verifyReleaseManifest(
  value: string | Buffer | Record<string, unknown>,
  trustedKeys: ReadonlyMap<string, Uint8Array> | Record<string, Uint8Array>,
): ReleaseManifest {
  let document: Record<string, unknown>;
  try {
    const parsed: unknown = typeof value === "object" && !Buffer.isBuffer(value)
      ? value
      : JSON.parse(Buffer.isBuffer(value) ? value.toString("utf8") : value);
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new TypeError();
    document = { ...(parsed as Record<string, unknown>) };
  } catch {
    fail("Manifest is not valid JSON", "manifest_invalid");
  }
  const signature = document.signature;
  if (!signature || typeof signature !== "object" || Array.isArray(signature)) {
    fail("Manifest signature block is invalid", "manifest_invalid_signature");
  }
  const signatureDocument = signature as Record<string, unknown>;
  if (Object.keys(signatureDocument).sort().join(",") !== "algorithm,keyId,value") {
    fail("Manifest signature block is invalid", "manifest_invalid_signature");
  }
  if (signatureDocument.algorithm !== "ed25519") {
    fail("Manifest signature algorithm is unsupported", "manifest_algorithm_unsupported");
  }
  const keyId = signatureDocument.keyId;
  if (typeof keyId !== "string") fail("Manifest signing key is unknown", "manifest_key_unknown");
  const keySource = trustedKeys instanceof Map
    ? trustedKeys.get(keyId)
    : (trustedKeys as Record<string, Uint8Array>)[keyId];
  if (!keySource || keySource.length !== 32) fail("Manifest signing key is unknown", "manifest_key_unknown");
  let signatureBytes: Buffer;
  try {
    signatureBytes = decodeBase64Url(signatureDocument.value, 64);
  } catch {
    fail("Manifest signature encoding is invalid", "manifest_invalid_signature");
  }
  const payload = { ...document };
  delete payload.signature;
  const publicKey = createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, Buffer.from(keySource)]),
    format: "der",
    type: "spki",
  });
  if (!verifySignature(null, canonicalJson(payload), publicKey, signatureBytes)) {
    fail("Manifest signature is invalid", "manifest_invalid_signature");
  }
  if (document.schemaVersion !== 1) fail("Manifest schema is unsupported", "manifest_schema_unsupported");
  if (typeof document.browserVersion !== "string" || !document.browserVersion) {
    fail("Manifest browser version is invalid", "manifest_invalid");
  }
  if (typeof document.sdkCompatibility !== "string" || !document.sdkCompatibility) {
    fail("Manifest SDK compatibility is invalid", "manifest_invalid");
  }
  if (document.status !== "available" && document.status !== "revoked") {
    fail("Manifest release status is invalid", "manifest_invalid");
  }
  if (!Array.isArray(document.artifacts) || document.artifacts.length === 0) {
    fail("Manifest artifacts are invalid", "manifest_invalid");
  }
  const artifacts = document.artifacts.map(parseArtifact);
  const identities = artifacts.map((item) => `${item.platform}/${item.arch}`);
  if (new Set(identities).size !== identities.length) {
    fail("Manifest has duplicate platform artifacts", "manifest_duplicate_artifact");
  }
  const evidence = document.evidence === undefined ? undefined : parseEvidence(document.evidence);
  return { ...(document as unknown as ReleaseManifest), artifacts, ...(evidence === undefined ? {} : { evidence }) };
}

function parseEvidenceArtifact(value: unknown, mediaType: ReleaseEvidenceArtifact["mediaType"]): ReleaseEvidenceArtifact {
  if (!value || typeof value !== "object" || Array.isArray(value)) fail("Release evidence is invalid", "manifest_evidence_invalid");
  const document = value as Record<string, unknown>;
  if (Object.keys(document).sort().join(",") !== "mediaType,sha256,size,url" ||
      typeof document.url !== "string" || !document.url.startsWith("https://") ||
      typeof document.sha256 !== "string" || !/^[a-f0-9]{64}$/.test(document.sha256) ||
      !Number.isSafeInteger(document.size) || (document.size as number) <= 0 || document.mediaType !== mediaType) {
    fail("Release evidence is invalid", "manifest_evidence_invalid");
  }
  return document as unknown as ReleaseEvidenceArtifact;
}

function parseEvidence(value: unknown): ReleaseEvidence {
  if (!value || typeof value !== "object" || Array.isArray(value)) fail("Release evidence is missing", "manifest_evidence_missing");
  const document = value as Record<string, unknown>;
  if (Object.keys(document).sort().join(",") !== "chromiumPatchInventory,provenance,sbom,sourceBoundary") {
    fail("Release evidence fields are invalid", "manifest_evidence_invalid");
  }
  const boundary = document.sourceBoundary;
  if (!boundary || typeof boundary !== "object" || Array.isArray(boundary)) fail("Source boundary is invalid", "manifest_evidence_invalid");
  const sourceBoundary = boundary as Record<string, unknown>;
  if (Object.keys(sourceBoundary).sort().join(",") !== "chromiumPatches,proprietaryCore,sdk" ||
      sourceBoundary.sdk !== "open-source" || sourceBoundary.chromiumPatches !== "inventory-and-approved-patches" ||
      sourceBoundary.proprietaryCore !== "private") {
    fail("Source boundary is invalid", "manifest_evidence_invalid");
  }
  return {
    sbom: parseEvidenceArtifact(document.sbom, "application/vnd.cyclonedx+json"),
    provenance: parseEvidenceArtifact(document.provenance, "application/vnd.in-toto+json"),
    chromiumPatchInventory: parseEvidenceArtifact(
      document.chromiumPatchInventory,
      "application/vnd.slybrowser.chromium-patch-inventory+json",
    ),
    sourceBoundary: sourceBoundary as unknown as ReleaseEvidence["sourceBoundary"],
  };
}

function parseArtifact(value: unknown): ReleaseArtifact {
  if (!value || typeof value !== "object" || Array.isArray(value)) fail("Artifact is invalid", "manifest_invalid");
  const document = value as Record<string, unknown>;
  const keys = Object.keys(document);
  const requiredKeys = [
    "arch", "archiveFormat", "browserExecutable", "browserSha256", "driverExecutable",
    "driverSha256", "platform", "privateModules", "resources", "sha256", "size", "url",
  ];
  const allowedKeys = new Set([...requiredKeys, "codeSignature"]);
  if (requiredKeys.some((key) => !keys.includes(key)) || keys.some((key) => !allowedKeys.has(key))) {
    fail("Artifact fields are invalid", "manifest_invalid");
  }
  if (!new Set(["windows", "linux", "macos"]).has(document.platform as string)) {
    fail("Artifact platform is invalid", "manifest_invalid");
  }
  if (!new Set(["x64", "arm64"]).has(document.arch as string)) {
    fail("Artifact architecture is invalid", "manifest_invalid");
  }
  if (typeof document.url !== "string" || !document.url.startsWith("https://")) {
    fail("Artifact URL must use HTTPS", "manifest_invalid");
  }
  if (typeof document.sha256 !== "string" || !/^[a-f0-9]{64}$/.test(document.sha256)) {
    fail("Artifact SHA-256 is invalid", "manifest_invalid");
  }
  if (!Number.isSafeInteger(document.size) || (document.size as number) <= 0) {
    fail("Artifact size is invalid", "manifest_invalid");
  }
  if (document.archiveFormat !== "zip" ||
      typeof document.browserExecutable !== "string" || !safeRelativePath(document.browserExecutable) ||
      typeof document.driverExecutable !== "string" || !safeRelativePath(document.driverExecutable) ||
      typeof document.browserSha256 !== "string" || !/^[a-f0-9]{64}$/.test(document.browserSha256) ||
      typeof document.driverSha256 !== "string" || !/^[a-f0-9]{64}$/.test(document.driverSha256)) {
    fail("Artifact runtime metadata is invalid", "manifest_invalid");
  }
  const privateModules = parsePrivateModules(document.privateModules);
  const resources = parseResources(document.resources);
  const codeSignature = Object.hasOwn(document, "codeSignature")
    ? parseCodeSignature(document.codeSignature)
    : undefined;
  return {
    ...(document as unknown as ReleaseArtifact),
    privateModules,
    resources,
    ...(codeSignature === undefined ? {} : { codeSignature }),
  };
}

function safeRelativePath(value: string): boolean {
  return value.length > 0 && value.length <= 512 && !value.startsWith("/") && !value.startsWith("\\") &&
    !/^[A-Za-z]:/.test(value) && !value.replaceAll("\\", "/").split("/").includes("..");
}

function parsePrivateModules(value: unknown): ReleasePrivateModule[] {
  if (!Array.isArray(value) || value.length === 0) fail("Artifact private module metadata is invalid", "manifest_invalid");
  return value.map((item) => {
    if (!item || typeof item !== "object" || Array.isArray(item)) fail("Artifact private module metadata is invalid", "manifest_invalid");
    const document = item as Record<string, unknown>;
    if (Object.keys(document).sort().join(",") !== "abi,path,sha256,size" ||
        typeof document.path !== "string" || !safeRelativePath(document.path) ||
        typeof document.sha256 !== "string" || !/^[a-f0-9]{64}$/.test(document.sha256) ||
        !Number.isSafeInteger(document.size) || (document.size as number) <= 0 ||
        typeof document.abi !== "string" || document.abi.length === 0 || document.abi.length > 128) {
      fail("Artifact private module metadata is invalid", "manifest_invalid");
    }
    return document as unknown as ReleasePrivateModule;
  });
}

function parseResources(value: unknown): ReleaseResourceFile[] {
  if (!Array.isArray(value) || value.length === 0) fail("Artifact resource metadata is invalid", "manifest_invalid");
  return value.map((item) => {
    if (!item || typeof item !== "object" || Array.isArray(item)) fail("Artifact resource metadata is invalid", "manifest_invalid");
    const document = item as Record<string, unknown>;
    if (Object.keys(document).sort().join(",") !== "path,sha256,size" ||
        typeof document.path !== "string" || !safeRelativePath(document.path) ||
        typeof document.sha256 !== "string" || !/^[a-f0-9]{64}$/.test(document.sha256) ||
        !Number.isSafeInteger(document.size) || (document.size as number) <= 0) {
      fail("Artifact resource metadata is invalid", "manifest_invalid");
    }
    return document as unknown as ReleaseResourceFile;
  });
}

function parseCodeSignature(value: unknown): ReleaseCodeSignature {
  if (!value || typeof value !== "object" || Array.isArray(value)) fail("Artifact code signature metadata is invalid", "manifest_invalid");
  const document = value as Record<string, unknown>;
  if (Object.keys(document).sort().join(",") !== "certificateSha256,scheme,subject,timestampRequired" ||
      !new Set(["authenticode", "apple-developer-id", "x509-code-signing"]).has(document.scheme as string) ||
      typeof document.subject !== "string" || document.subject.length === 0 || document.subject.length > 512 ||
      typeof document.certificateSha256 !== "string" || !/^[a-f0-9]{64}$/.test(document.certificateSha256) ||
      typeof document.timestampRequired !== "boolean") {
    fail("Artifact code signature metadata is invalid", "manifest_invalid");
  }
  return document as unknown as ReleaseCodeSignature;
}

export function selectArtifact(manifest: ReleaseManifest, platform: string, arch: string): ReleaseArtifact {
  const matches = manifest.artifacts.filter((item) => item.platform === platform && item.arch === arch);
  if (matches.length !== 1) {
    const available = manifest.artifacts.map((item) => `${item.platform}/${item.arch}`).sort().join(", ");
    fail(`No signed artifact supports ${platform}/${arch}; available targets: ${available}`, "artifact_not_found");
  }
  return matches[0]!;
}

export async function verifyArtifact(path: string, artifact: ReleaseArtifact): Promise<void> {
  let info;
  try {
    info = await stat(path);
  } catch {
    throw new ArtifactError("Browser artifact is missing", "artifact_missing");
  }
  if (!info.isFile()) throw new ArtifactError("Browser artifact is missing", "artifact_missing");
  if (info.size !== artifact.size) throw new ArtifactError("Browser artifact size does not match", "artifact_size_mismatch");
  const digest = createHash("sha256");
  await new Promise<void>((resolve, reject) => {
    const stream = createReadStream(path);
    stream.on("data", (chunk) => digest.update(chunk));
    stream.on("error", reject);
    stream.on("end", resolve);
  });
  if (digest.digest("hex") !== artifact.sha256) {
    throw new ArtifactError("Browser artifact checksum does not match", "artifact_hash_mismatch");
  }
}
